File: JavaScript/186-Reverse Words in a String II.js
/**
 * @param {character[]} s
 * @return {void} Do not return anything, modify s in-place instead.
 */
var reverseWords = function(s) {
    const N = s.length;
    s.reverse();
    let start = 0;
    
    // Array.prototype.slice returns a copy already. Array.prototype.splice modifies the original array
    
    for (let end = 0; end < N; end++){
        if (s[end] == " "){
            s.splice(start, end-start, ...s.slice(start, end).reverse() );
            start = end + 1;
        }
    }
    s.splice(start, N-start, ...s.slice(start, N).reverse());
};

File: 186-Reverse Words in a String II.py
class Solution(object):
    def reverseWords(self, s):
        """
        :type s: List[str]
        :rtype: None Do not return anything, modify s in-place instead.
        """
        N = len(s)
        s[:] = s[::-1]
        start = 0
        
        for end in range(N):
            if s[end] == " ":
                s[start:end] = s[start:end][::-1]
                start = end + 1
        
        s[start:N] = s[start:N][::-1]